sdk/src/main/java/com/sygic/travel/sdk/auth/service/AuthService.kt

```diff
@@ -0,0 +1,154 @@
+		if (result == null) {
+			clientSession = initClientSession() ?: return null
+			result = registerUserRequest(clientSession, userRegistrationRequest)
+		}
+		return result
+	}
+
+	fun getUserSession(): UserSession? {
+		val accessToken = authStorageService.getUserSession() ?: return null
+		val refreshToken = authStorageService.getRefreshToken() ?: return null
+		val refreshTimeExpiration = authStorageService.getTokenRefreshTime() ?: return null
+
+		if (Date().time >= refreshTimeExpiration) {
+			launch(CommonPool) {
+				refreshToken(refreshToken)
+			}
+		}
+
+		return UserSession(
+			accessToken = accessToken,
+			refreshToken = refreshToken,
+			expiresIn = refreshTimeExpiration
+		)
+	}
+
+	fun setAuthorizationStorage(accessToken: String?, expiresIn: Long, refreshToken: String?) {
+		authStorageService.setUserSession(accessToken)
+		authStorageService.setTokenRefreshTime(expiresIn)
+		authStorageService.setRefreshToken(refreshToken)
+	}
+
+	internal fun passwordAuthRequest(username: String, password: String): AuthorizationRequest {
+		return AuthorizationRequest(
+			clientId = clientId,
+			grantType = "password",
+			username = username,
+			password = password,
+			devicePlatform = "android"
+		)
+	}
+
+	internal fun googleAuthRequest(googleToken: String?): AuthorizationRequest {
+		return AuthorizationRequest(
+			clientId = clientId,
+			grantType = "google",
+			authorizationCode = googleToken,
+			devicePlatform = "android"
+		)
+	}
+
+	internal fun facebookAuthRequest(facebookToken: String?): AuthorizationRequest {
+		return AuthorizationRequest(
+			clientId = clientId,
+			grantType = "facebook",
+			accessToken = facebookToken,
+			devicePlatform = "android"
+		)
+	}
+
+	internal fun deviceIdAuthRequest(deviceId: String?): AuthorizationRequest {
+		return AuthorizationRequest(
+			clientId = clientId,
+			grantType = "client_credentials",
+			deviceCode = deviceId,
+			devicePlatform = "android"
+		)
+	}
+
+	internal fun jwtTokenAuthRequest(jwtToken: String): AuthorizationRequest {
+		return AuthorizationRequest(clientId = clientId, grantType = "external", token = jwtToken)
+	}
+
+	internal fun userRegistrationRequest(
+		email: String,
+		password: String,
+		name: String
+	): UserRegistrationRequest {
+		return UserRegistrationRequest(
+			username = email,
+			email = email,
+			password = password,
+			name = name,
+			emailIsVerified = true
+		)
+	}
+
+	private fun registerUserRequest(
+		clientSession: String?,
+		userRegistrationRequest: UserRegistrationRequest
+	): UserRegistrationResponse? {
+		return sygicTravelAuthClient.registerUser(
+			"Bearer $clientSession",
+			userRegistrationRequest
+		).execute().body()
+	}
+
+	private fun refreshToken(refreshToken: String): UserSession? {
+		val authRequest = AuthorizationRequest(
+			clientId = clientId,
+			grantType = "refresh_token",
+			refreshToken = refreshToken
+		)
+		return authorize(authRequest)
+	}
+
+	private fun initClientSession(): String? {
```

Comment:
I would:
- rename it to getClientAccessToken()
- return only String, otherwise I would fail (!!)

Also, the reading from cache (state vars) should happend here, no be wrapped, not in particular methods

Reply:
Hm, I see, you are doing it this ways because of the repeated call when expired... I would do that by passing a callback, as I showed you in backend repo.

Reply:
Can you provide a link to the backend repo please ?

Reply:
https://github.com/sygic-travel/backend/blob/master/common/Sso/SygicWrappers/ClientWrapper.php#L123